x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/task/OpenJobPersistentTasksExecutor.java

```diff
@@ -454,51 +460,92 @@ private void verifyCurrentSnapshotVersion(String jobId, ActionListener<Boolean>
+                    request.setSkipJobStateValidation(true);
+                    request.masterNodeTimeout(PERSISTENT_TASK_MASTER_NODE_TIMEOUT);
+                    request.timeout(PERSISTENT_TASK_MASTER_NODE_TIMEOUT);
+                    executeAsyncWithOrigin(
+                        client,
+                        ML_ORIGIN,
+                        ResetJobAction.INSTANCE,
+                        request,
+                        ActionListener.wrap(response -> listener.onResponse(true), listener::onFailure)
+                    );
+                } else {
+                    logger.info("[{}] job has running datafeed task; reverting to current snapshot", jobTask.getJobId());
+                    RevertModelSnapshotAction.Request request = new RevertModelSnapshotAction.Request(
+                        jobTask.getJobId(),
+                        jobSnapshotId == null ? ModelSnapshot.EMPTY_SNAPSHOT_ID : jobSnapshotId
+                    );
+                    request.setForce(true);
+                    request.setDeleteInterveningResults(true);
+                    request.masterNodeTimeout(PERSISTENT_TASK_MASTER_NODE_TIMEOUT);
+                    request.timeout(PERSISTENT_TASK_MASTER_NODE_TIMEOUT);
+                    executeAsyncWithOrigin(
+                        client,
+                        ML_ORIGIN,
+                        RevertModelSnapshotAction.INSTANCE,
+                        request,
+                        ActionListener.wrap(response -> listener.onResponse(true), listener::onFailure)
+                    );
+                }
+            }, error -> listener.onFailure(ExceptionsHelper.serverError("[{}] error getting job", error, jobTask.getJobId())));
 
-        // We need to refetch the job in order to learn what is its current model snapshot
-        // as the one that exists in the task params is outdated.
-        GetJobsAction.Request request = new GetJobsAction.Request(jobId);
-        request.masterNodeTimeout(PERSISTENT_TASK_MASTER_NODE_TIMEOUT);
-        executeAsyncWithOrigin(client, ML_ORIGIN, GetJobsAction.INSTANCE, request, jobListener);
+            // We need to refetch the job in order to learn what is its current model snapshot
+            // as the one that exists in the task params is outdated.
+            GetJobsAction.Request request = new GetJobsAction.Request(jobTask.getJobId());
+            request.masterNodeTimeout(PERSISTENT_TASK_MASTER_NODE_TIMEOUT);
+            executeAsyncWithOrigin(client, ML_ORIGIN, GetJobsAction.INSTANCE, request, jobListener);
+        }
+
+        @Override
+        public boolean shouldRetry(Exception e) {
+            if (jobTask.isClosing() || jobTask.isVacating()) {
+                return false;
+            }
+            logger.error(
+                new ParameterizedMessage("[{}] error reverting job to its current snapshot; attempting retry", jobTask.getJobId()),
+                e
+            );
+            return true;
```

Comment:
Part of me thinks we will run into a scenario where we retry TOO much. If thats the case, say we retry on an unrecoverable scenario,

starting at 1 millisecond we will retry around 19 more times (`2^20 > 900000`). Which is 19 logs all saying the same thing. 

I do think we should log here, but I am not sure it should be every time. Because in a catastrophe, we will have 19 logs per job which could drown out any other signal.